### ISSUE TITLE
opentitan on Genesys 2

### ISSUE THREAD

**Issue by jackieke724** [lowRISC/opentitan#4640]
Hello,

I am trying to implement earlgrey on [Genesys 2](https://reference.digilentinc.com/reference/programmable-logic/genesys-2/start?redirect=1) which is a similar SOC as Nexys Video, but I am not seeing any outputs through UART or LEDs.

So far, I have changed the pin locations of peripherals and IO_clock accordingly. For the clock, I have added an `IBUFGDS` due to the 200 MHz differential clock pins on the board. There are other clock-related changes in `clkgen` and clocks.xdc. I have disabled placement.xdc and the opt_design_post tcl hook (I am actually not sure what they are for in the first place). 

The bitstream was completed with no obvious warnings to me. As I checked with the unchanged project, there were a few additional warnings about `Sequential element (...) is unused and will be removed`.

I was suspecting that the boot_rom.vmem file did not run properly on my board. I added the following code to assign LEDs to a fixed value though GPIO. When I baked that vmem into the bitstream, I still saw no outputs from LEDs.
```
  dif_gpio_params_t gpio_params = {
	.base_addr = mmio_region_from_addr(TOP_EARLGREY_GPIO_BASE_ADDR),
  };
  CHECK(dif_gpio_init(gpio_params, &gpio) == kDifGpioOk);
  // Enable GPIO: 0-7 and 16 is input; 8-15 is output.
  CHECK(dif_gpio_output_set_enabled_all(&gpio, 0xff00) == kDifGpioOk);
  CHECK(dif_gpio_write_all(&gpio, 0x0f00) == kDifGpioOk);
```

Another problem I have is loading the hello world app. When I used `build-bin/sw/host/spiflash/spiflash --input build-bin/sw/device/examples/hello_world/hello_world_fpga_nexysvideo.bin`, I got:
> Unable to open FTDI SPI interface.

The hardware ID of the JTAG COM port is shown below. How can I find its dev-id and dev-sn?
> ACPI\VEN_PNP&DEV_0501
> ACPI\PNP0501
> *PNP0501

Thank you for any insights!

**Comment by jackieke724** [comment]
Do you have any tips for debugging? How can I find out where the boot_rom code got stuck?

**Comment by tjaychen** [comment]
just to double check, after you changed the clock, did you also make appropriate changes to the boot code to reflect the clock change? Otherwise the baud rates could be way off ?

For the spi flash FTDI issues, I'm not really an expert, maybe others can chime in.

**Comment by jackieke724** [comment]
@tjaychen ,thanks for the inputs. Could you be more specific on changes required by the boot code?
For RTL, it is just the `IO_CLK` frequency is changed to 200 MHz. It is fed to the PLL to generate the CPU and peripheral clocks which are still 10 MHz and 48 MHz, repectively. In this scenario, what other changes do I need?

**Comment by jackieke724** [comment]
I have made a little progress today by setting up OpenOCD and GDB. The initial output from GDB is shown below:
> GNU gdb (crosstool-NG 1.24.0.41-475d72f) 8.2.1                                                                  
> Copyright (C) 2018 Free Software Foundation, Inc.                                                               
> License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>                                   
> This is free software: you are free to change and redistribute it.                                              
> There is NO WARRANTY, to the extent permitted by law.                                                           
> Type "show copying" and "show warranty" for details.                                                            
> This GDB was configured as "--host=x86_64-build_pc-linux-gnu --target=riscv32-unknown-elf".                     
> Type "show configuration" for configuration details.                                                            
> For bug reporting instructions, please see:                                                                     
> <toolchains@lowrisc.org>.                                                                                       
> Find the GDB manual and other documentation resources online at:                                                
>     <http://www.gnu.org/software/gdb/documentation/>.                                                           
>                                                                                                                 
> For help, type "help".                                                                                          
> Type "apropos word" to search for commands related to "word"...                                                 
> Reading symbols from build-bin/sw/device/boot_rom/boot_rom_fpga_nexysvideo.elf...done.                          
> The target architecture is assumed to be riscv:rv32                                                             
> Remote debugging using :3333                                                                                    
> warning: Target-supplied registers are not supported by the current architecture                                
> 0x00008900 in ibex_mcycle_read () at /mnt/d/ubuntu_workspace/genesys2/opentitan/sw/device/lib/runtime/ibex.h:32 
> 32        asm volatile(                                                                                         
> ra             0x8296   0x8296 <_boot_start+342>                                                                
> sp             0x1000ffd0       0x1000ffd0                                                                      
> gp             0x10000800       0x10000800                                                                      
> tp             0x0      0x0                                                                                     
> t0             0x4      4                                                                                       
> t1             0x98ac8000       2561441792                                                                      
> t2             0xcccccccd       3435973837                                                                      
> s0             0x1000ffd8       0x1000ffd8                                                                      
> s1             0xa744   42820                                                                                   
> a0             0x0      0                                                                                       
> a1             0xf00    3840                                                                                    
> a2             0x40010000       1073807360                                                                      
> a3             0x0      0                                                                                       
> a4             0xac1d   44061                                                                                   
> a5             0xa8ff   43263                                                                                   
> a6             0x1000ff0b       268500747                                                                       
> a7             0x9896   39062                                                                                   
> s2             0x0      0                                                                                       
> s3             0x0      0                                                                                       
> s4             0x0      0                                                                                       
> s5             0x0      0                                                                                       
> s6             0x0      0                                                                                       
> s7             0x0      0                                                                                       
> s8             0x0      0                                                                                       
> s9             0x0      0                                                                                       
> s10            0x0      0                                                                                       
> s11            0x0      0                                                                                       
> t3             0x18     24                                                                                      
> t4             0x8a1    2209                                                                                    
> --Type <RET> for more, q to quit, c to continue without paging--                                                
> t5             0x0      0                                                                                       
> t6             0x0      0                                                                                       
> pc             0x8900   0x8900 <usleep+66>      

Questions:
1. I am still just learning the system and Ibex. Could any of you share some tips for debugging? How would I find out where the boot code hanged?
2. Any ideas about why FTDI SPI interface uable to be opened?

Any input is appreciated!

                                                               
_Aside for those who are working in **WSL** ( which does not support lsusb yet):_
I installed OpenOCD on Windows following this [tutorial](http://pansila.github.io/posts/7db4884d/). Then following the [getting_started_fpga guide](https://docs.opentitan.org/doc/ug/getting_started_fpga/index.html) up to _Debug with gdb_, OpenOCD kept dropping GDB connections after every connect. I had to use the command below:
```
/tools/riscv/bin/riscv32-unknown-elf-gdb -ex "set architecture riscv:rv32" -ex "target extended-remote :3333" -ex "info reg" build-bin/sw/device/boot_rom/boot_rom_fpga_nexysvideo.elf
```

**Comment by tjaychen** [comment]
sorry i'm not familiar enough with openocd / gdb to give you a ton of input there.. but @imphil may be able to spot something.
For the device Id and device serial number, just trying using lsusb.  You can also find some info in the `README.md` file under `sw/host/spiflash`

Regarding the clock change, do you mean to say you just changed the input clock frequency? Did you also replace this [file](https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/rtl/clkgen_xil7series.sv) with something that is appropriate for your FPGA? If all the parameters are kept the not same, I'm not sure if that means you end up generating a clock that is twice as fast?

**Comment by jackieke724** [comment]
@tjaychen thank you for your reply during the break. Both Nexys Video and Genesys 2 use the FT2232 chip, so I guess it should be fine to keep the device ID and serial number inputs as the default. lsusb is not supported in WSL, but here is a screenshot of USBDeview. I have tried using the dev-id and dev-sn with the numbers in the picture, as well as just the default.
![image](https://user-images.githubusercontent.com/15860703/103255863-f142bd80-49c5-11eb-96b1-6e9df7e54523.png)

For the clock change, I did change the clkgen module you pointed. To be specific, I will list the git diff results below.
hw/top_earlgrey/data/**clocks.xdc**
```
diff --git a/hw/top_earlgrey/data/clocks.xdc b/hw/top_earlgrey/data/clocks.xdc
index 3d67989a..035bfadd 100644
--- a/hw/top_earlgrey/data/clocks.xdc
+++ b/hw/top_earlgrey/data/clocks.xdc
@@ -3,7 +3,7 @@
 ## SPDX-License-Identifier: Apache-2.0

 ## Clock Signal
-create_clock -add -name sys_clk_pin -period 10.00 -waveform {0 5} [get_ports IO_CLK]
+create_clock -add -name sys_clk_pin -period 5.00 -waveform {0 2.5} [get_ports SYSCLK_P]

 ## Clock Domain Crossings
 set clks_10_unbuf [get_clocks -of_objects [get_pin clkgen/pll/CLKOUT0]]
```

hw/top_earlgrey/rtl/**clkgen_xil7series.sv**
```
diff --git a/hw/top_earlgrey/rtl/clkgen_xil7series.sv b/hw/top_earlgrey/rtl/clkgen_xil7series.sv
index 2822a62b..a99492b8 100644
--- a/hw/top_earlgrey/rtl/clkgen_xil7series.sv
+++ b/hw/top_earlgrey/rtl/clkgen_xil7series.sv
@@ -40,7 +40,7 @@ module clkgen_xil7series # (
     .COMPENSATION         ("ZHOLD"),
     .STARTUP_WAIT         ("FALSE"),
     .DIVCLK_DIVIDE        (1),
-    .CLKFBOUT_MULT        (12),
+    .CLKFBOUT_MULT        (6),
     .CLKFBOUT_PHASE       (0.000),
     .CLKOUT0_DIVIDE       (120),
     .CLKOUT0_PHASE        (0.000),
@@ -48,7 +48,7 @@ module clkgen_xil7series # (
     .CLKOUT1_DIVIDE       (25),
     .CLKOUT1_PHASE        (0.000),
     .CLKOUT1_DUTY_CYCLE   (0.500),
-    .CLKIN1_PERIOD        (10.000)
+    .CLKIN1_PERIOD        (5.000)
   ) pll (
     .CLKFBOUT            (clk_fb_unbuf),
     .CLKOUT0             (clk_10_unbuf),
```

hw/top_earlgrey/rtl/**top_earlgrey_nexysvideo.sv**
```
diff --git a/hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv b/hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv
index 7b07ce22..4c767c3c 100644
--- a/hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv
+++ b/hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv
@@ -8,7 +8,8 @@ module top_earlgrey_nexysvideo #(
   parameter BootRomInitFile = "boot_rom_fpga_nexysvideo.32.vmem"
 ) (
   // Clock and Reset
-  input               IO_CLK,
+       input               SYSCLK_N,
+       input               SYSCLK_P,
   input               IO_RST_N,
   // JTAG interface
   inout               IO_DPS0, // IO_JTCK,    IO_SDCK
@@ -57,6 +58,22 @@ module top_earlgrey_nexysvideo #(
 );

   import top_earlgrey_pkg::*;
+
+       //////////////////////
+  // LVDS Clock                        //
+  //////////////////////
+
+       logic IO_CLK;
+       IBUFGDS #(
+               .DIFF_TERM("FALSE"), // Differential Termination
+               .IBUF_LOW_PWR("TRUE"), // Low power="TRUE", Highest performance="FALSE"
+               .IOSTANDARD("DEFAULT") // Specify the input I/O standard
+               ) IBUFGDS_inst (
+               .O(IO_CLK), // Clock buffer output
+               .I(SYSCLK_P), // Diff_p clock buffer input (connect directly to top-level port)
+               .IB(SYSCLK_N) // Diff_n clock buffer input (connect directly to top-level port)
+       );
+

   //////////////////////
   // Padring Instance //
```

**Comment by jackieke724** [comment]
I have made very little debugging progress. To summarize, the boot rom code runs fine. I am able to receive button inputs and output them to leds. The following 2 problems persist.

1. No uart outputs on the serial console
2. cannot open FTDI SPI interface

I have made uart working with Microblaze and AXI Uartlite, so the board and the cables are okay. The serial console setting should be the same as in the doc. I have also tried to compile on Vivado 2020.2, 2020.1, and 2018.3 on both Windows and Linux.

Could anyone please shed some light on these issues?

**Comment by imphil** [comment]
@jackieke724 Porting to a new FPGA board is never trivial and requires a bit of experience. To be able to help (without having the board ourselves), could you please:

1. Publish your changes in a git repository somewhere. Please ensure that you have reasonably clean commits on what you changed, otherwise debugging is  impossible.
4. Windows/WSL adds another layer of complication to the debugging story. To make our life easier please use a native Linux machine for the first bringup.
2. Start from the official XDC file by Digilent for the board. Adjust it to match the OpenTitan top-level pin names, and carefully double-check each pin assignment.
3. You don't need to use spiflash to check if UART is working, as soon as you flash the bitstream, connect over UART, and press the "CPU reset" button you should see the bootrom message in the UART output.
5. Check in Vivado that the derived frequencies are correct (i.e. identical to the ones in the Nexys Video design)
6. Add a ILA debugging core in Vivado to the design and check if you see activity on the UART outputs.

**Comment by jackieke724** [comment]
Hello @imphil @tjaychen , thank you for your suggestions. I have gone through steps 1-6 and oscilloscope as you suggested. At the end, I finally resolved the UART and spiflash issues.

For the UART, I had the pin assignments swapped. I basically overlooked the "rx_out" and "tx_in" net names used in the official XDC file, and mis-interpreted them as equal to IO_URX and IO_UTX respectively. Sorry for bringing up the silly mistake.

For the spiflash, I was able to use it on a virtual machine. I think the lack of lsusb support of WSL currently was the problem.
